contracts/utils/OrderPayable.sol

```diff
@@ -0,0 +1,142 @@
+
+    /// @notice Withdrawer role
+    bytes32 public immutable override withdrawerRole;
+
+    /// @notice Returns if the order with ID is paid for
+    mapping(bytes32 => bool) public override orderIdToPaymentStatus;
+
+    /// @param _accessControlRegistry AccessControlRegistry contract address
+    /// @param _adminRoleDescription Admin role description
+    /// @param _manager Manager address
+    constructor(
+        address _accessControlRegistry,
+        string memory _adminRoleDescription,
+        address _manager
+    )
+        AccessControlRegistryAdminnedWithManager(
+            _accessControlRegistry,
+            _adminRoleDescription,
+            _manager
+        )
+    {
+        orderSignerRole = _deriveRole(
+            _deriveAdminRole(manager),
+            ORDER_SIGNER_ROLE_DESCRIPTION
+        );
+        withdrawerRole = _deriveRole(
+            _deriveAdminRole(manager),
+            WITHDRAWER_ROLE_DESCRIPTION
+        );
+    }
+
+    /// @notice Called with value to pay for an order
+    /// @dev The sender must set `msg.value` to cover the exact amount
+    /// specified by the order.
+    /// Input arguments are provided in encoded form to improve the UX for
+    /// using ABI-based, automatically generated contract GUIs such as ones
+    /// from Safe and Etherscan. Given that OrderPayable is verified, the user
+    /// is only required to provide the OrderPayable address, select
+    /// `payForOrder()`, copy-paste `encodedData` (instead of 4 separate
+    /// fields) and enter `msg.value`.
+    /// @param encodedData The order ID, expiration timestamp, order signer
+    /// address and signature in ABI-encoded form
+    function payForOrder(bytes calldata encodedData) external payable override {
+        // Do not care if `encodedData` has trailing data
+        (
+            bytes32 orderId,
+            uint256 expirationTimestamp,
+            address orderSigner,
+            bytes memory signature
+        ) = abi.decode(encodedData, (bytes32, uint256, address, bytes));
+        // We do not allow invalid orders even if they are signed by an
+        // authorized order signer
+        require(orderId != bytes32(0), "Order ID zero");
+        require(expirationTimestamp > block.timestamp, "Order expired");
+        require(
+            orderSigner == manager ||
+                IAccessControlRegistry(accessControlRegistry).hasRole(
+                    orderSignerRole,
+                    orderSigner
+                ),
+            "Invalid order signer"
+        );
+        require(msg.value > 0, "Payment amount zero");
+        require(!orderIdToPaymentStatus[orderId], "Order already paid for");
+        require(
+            (
+                keccak256(
+                    abi.encodePacked(
+                        block.chainid,
+                        address(this),
+                        orderId,
+                        expirationTimestamp,
```

Comment:
shouldn't we add `msg.sender` to this signature. I know we will have a check in the backend but I don't think anyone should be able to execute orders if it's signed by the backend.